packages/trie/src/util/asyncWalk.ts

```diff
@@ -0,0 +1,60 @@
+import { RLP } from '@ethereumjs/rlp'
+import { equalsBytes, toHex } from 'ethereum-cryptography/utils'
+
+import { BranchNode } from '../node/branch.js'
+import { ExtensionNode } from '../node/extension.js'
+
+import type { Trie } from '../trie'
+import type { TrieNode } from '../types'
+
+export type NodeFilter = (node: TrieNode, key: number[]) => Promise<boolean>
+export type OnFound = (node: TrieNode, key: number[]) => Promise<any>
+
+/**
+ * Walk Trie via async generator
+ * @param nodeHash - The root key to walk on.
+ * @param currentKey - The current (partial) key.
+ * @param onFound - Called on every node found (before filter)
+ * @param filter - Filter nodes yielded by the generator.
+ * @param visited - Set of visited nodes
+ * @returns AsyncIterable<{ node: TrieNode; currentKey: number[] }>
+ * Iterate through nodes with
+ * `for await (const { node, currentKey } of trie._walkTrie(root)) { ... }`
+ */
+export async function* _walkTrie(
+  this: Trie,
+  nodeHash: Uint8Array,
+  currentKey: number[] = [],
+  onFound: OnFound = async (_trieNode: TrieNode, _key: number[]) => {},
+  filter: NodeFilter = async (_trieNode: TrieNode, _key: number[]) => true,
+  visited: Set<string> = new Set<string>()
+): AsyncIterable<{ node: TrieNode; currentKey: number[] }> {
+  if (equalsBytes(nodeHash, this.EMPTY_TRIE_ROOT)) {
+    return
+  }
+  try {
+    const node = await this.lookupNode(nodeHash)
+    if (node === undefined || visited.has(toHex(this.hash(node!.serialize())))) {
+      return
+    }
+    visited.add(toHex(this.hash(node!.serialize())))
```

Comment:
From working with trie.ts, I think it'd be awesome if there were a canonical representation of the trie's and node's hash values and if trie.ts would provide this serialization of values directly instead of having "users" implement their own thing.